esp/services/ws_smc/ws_smcService.hpp

```diff
@@ -154,15 +157,67 @@ class CActivityInfo : public CInterface, implements IInterface
     inline IArrayOf<IEspActiveWorkunit>& queryActiveWUs() { return aws; };
     inline IArrayOf<IEspServerJobQueue>& queryServerJobQueues() { return serverJobQueues; };
     inline IArrayOf<IEspDFUJob>& queryDFURecoveryJobs() { return DFURecoveryJobs; };
+    inline StringBuffer& queryTimeCached(StringBuffer& str) { return timeCached.getString(str, true); }
+};
+
+class CWsSMCEx;
+
+class CActivityInfoReader : public CInterface, implements IThreaded
+{
+    bool stopping = false;
+    bool detached = false;
+    unsigned autoRebuildSeconds = DEFAULTACTIVITYINFOCACHEAUTOREBUILDSECOND;
+    unsigned forceRebuildSeconds = DEFAULTACTIVITYINFOCACHEFORCEBUILDSECOND;
+    Owned<CActivityInfo> activityInfoCache;
+    Semaphore sem;
+    CriticalSection crit;
+    CThreaded threaded;
+    std::atomic<bool> waiting = {false};
+public:
+    CActivityInfoReader(unsigned _autoRebuildSeconds, unsigned _forceRebuildSeconds)
+        : autoRebuildSeconds(_autoRebuildSeconds), forceRebuildSeconds(_forceRebuildSeconds), threaded("ActivityInfoReader")
+    {
+        threaded.init(this);
+    };
+
+    ~CActivityInfoReader()
+    {
+        stopping = true;
+        sem.signal();
+        threaded.join();
+    }
+
+    virtual void threadmain() override;
+    CActivityInfo* getActivityInfo()
+    {
+        CriticalBlock b(crit);
+        assert(activityInfoCache);
+        if (!activityInfoCache->isCachedActivityInfoValid(forceRebuildSeconds))
+            rebuild();
+        return activityInfoCache.getLink();
+    }
+    void rebuild()
+    {
+        bool expected = true;
+        if (waiting.compare_exchange_strong(expected, false))
+            sem.signal();
+    }
+    void setDetachedState(bool _detached)
+    {
+        if (detached != _detached)
+        {
+            detached = _detached;
+            if (!detached)
+                rebuild();
```

Comment:
I don't think there's any point in signalling now..?
It it's blocked - then it will timeout and the thread will see !detached.
It it's not blocked - there's nothing to do.
So I think this code should simply set detached to false.

Reply:
yes I misread it as calling when detached which would be pointless.
If was detached and now isn't, yes sensible to signal.

If this can be called from multiple threads, it's not thread safe though.

Reply:
Should I add a new CriticalSection before line 209 to make it thread safe?